### PR TITLE
[Beam-6696] GroupIntoBatches transform for Python SDK

### PR DESCRIPTION
**Pull request by raheelkhan**
This PR adds GroupIntoBatches transform in the Python SDK

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by raheelkhan** [comment]
@aaltay

**Comment by raheelkhan** [comment]
@robertwb

File: sdks/python/apache_beam/transforms/util.py [hunk 1]
```diff
@@ -671,6 +679,60 @@ def WithKeys(pcoll, k):
   return pcoll | Map(lambda v: (k, v))
 
 
+@typehints.with_input_types(typehints.KV[K, V])
+class GroupIntoBatches(PTransform):
+  def __init__(self, batch_size):
+    self.batch_size = batch_size
+
+  @staticmethod
+  def of_size(batch_size):
```
Comment (aaltay) [inline sdks/python/apache_beam/transforms/util.py]: I do not believe we need `of_size` in Python. Users can directly use GroupIntoBatches().

File: sdks/python/apache_beam/transforms/util.py [hunk 2]
```diff
@@ -671,6 +679,60 @@ def WithKeys(pcoll, k):
...
+    return GroupIntoBatches(batch_size)
+
+  def expand(self, pcoll):
+    input_coder = coders.registry.get_coder(pcoll)
+    if not input_coder.is_kv_coder():
```
Comment (aaltay) [inline sdks/python/apache_beam/transforms/util.py]: Do you need to check this? Type hint should ensure that this is of kv type.

File: sdks/python/apache_beam/transforms/util.py [hunk 3]
```diff
@@ -671,6 +679,60 @@ def WithKeys(pcoll, k):
...
+
+
+def _pardo_group_into_batches(batch_size, input_coder):
+  ELEMENT_STATE = BagStateSpec('values', input_coder)
+  COUNT_STATE = CombiningValueStateSpec('count', input_coder, CountCombineFn())
+  EXPIRY_TIMER = TimerSpec('expiry', TimeDomain.WATERMARK)
+
+  class _GroupIntoBatchesDoFn(DoFn):
+
+    def process(self, element,
```
Comment (aaltay) [inline sdks/python/apache_beam/transforms/util.py]: State/Timers does not work for all runners in python.

Instead, see this alternative implementation (https://github.com/apache/beam/blob/11aacd2ea2e2b1fec54efc0557e016e2b7f8d944/sdks/python/apache_beam/transforms/util.py#L430) Similar to the `_WindowAwareBatchingDoFn`, you batch elements in process and emit in process and finish_bundle.

**Comment by aaltay** [comment]
I will be out of office next week, adding @udim as a reviewer to follow up.

File: sdks/python/apache_beam/transforms/util.py (hunk 3 above)
Comment (raheelkhan) [reply sdks/python/apache_beam/transforms/util.py]: I'll take a look into this, but the ticket [https://issues.apache.org/jira/browse/BEAM-6696](https://issues.apache.org/jira/browse/BEAM-6696) says that it should use state to batch across bundles.

How should I interpret this ?

Thanks

Comment (aaltay) [reply sdks/python/apache_beam/transforms/util.py]: Good point.

We can use state/timers as you did here. Mark it as experimental and add a strongly worded comment or maybe even an actual warning to say that use of this requires state/timer support from the runner.

File: sdks/python/apache_beam/transforms/util.py (hunk 1 above)
Comment (raheelkhan) [reply sdks/python/apache_beam/transforms/util.py]: fixed

File: sdks/python/apache_beam/transforms/util.py (hunk 2 above)
Comment (raheelkhan) [reply sdks/python/apache_beam/transforms/util.py]: fixed

File: sdks/python/apache_beam/transforms/util.py (hunk 3 above)
Comment (raheelkhan) [reply sdks/python/apache_beam/transforms/util.py]: fixed

Comment (aaltay) [reply sdks/python/apache_beam/transforms/util.py]: Thank you, resolving this.

File: sdks/python/apache_beam/transforms/util.py [hunk 4]
```diff
@@ -671,6 +681,56 @@ def WithKeys(pcoll, k):
   return pcoll | Map(lambda v: (k, v))
 
 
+@experimental()
+@typehints.with_input_types(typehints.KV[K, V])
+class GroupIntoBatches(PTransform):
```
Comment (aaltay) [inline sdks/python/apache_beam/transforms/util.py]: Could you add a pydocs comment to this transform. With, what does it do, what is its expectations, and mentioning that its experimental status?

Comment (raheelkhan) [reply sdks/python/apache_beam/transforms/util.py]: Done.

**Comment by robertwb** [comment]
This seems highly redundant with the existing BatchElements; it seems better to update the latter to work well in streaming mode rather than introduce a new transform.

**Comment by aaltay** [comment]
Difference as I understand is BatchElements will batch based on its estimator, works without state and timer APIs. GroupIntoBatches will batch into fixed size, works for streaming and matches the behavior of Java transform with same name.

I agree there is duplication. What is your suggestion to merging the two?

**Comment by robertwb** [comment]
I suggest we update BatchElements to be backed by state, assuming we can do
so without performance degredation. (The estimator is equivalent to fixed
size when min == max.) It would also likely make sense to add auto-sizing
capabilities to Java.

(Another difference is that GroupIntoBatches requires keyed input, and
batches per (and then drops) the key. If we keep the key, we should
probably still emit it as we do for Java.)

**Comment by aaltay** [comment]
That makes sense, but how would that work for runners that does not support state.

I moved your comment to https://issues.apache.org/jira/browse/BEAM-6696 so that we do not lose track here in this PR. (I will be OOO for a few weeks.)

**Pull request merged**